### DISCUSSION TITLE
failed to build with android ndk

### DISCUSSION THREAD

**Discussion by mokhtarabadi** [eycorsican/leaf#52] [General]
can't build the leaf-bin for android by using ndk (also cargo ndk), below the output logs from terminal:

```
   Compiling leaf v0.1.2 (/home/mohammad/Develop/Works/leaf/leaf)
error[E0425]: cannot find value `ENDPOINT_TCP_CONCURRENCY` in module `option`
  --> leaf/src/app/dispatcher.rs:91:72
   |
91 |         let (endpoint_tcp_tx, endpoint_tcp_rx) = mpsc::channel(option::ENDPOINT_TCP_CONCURRENCY);
   |                                                                        ^^^^^^^^^^^^^^^^^^^^^^^^ not found in `option`

error[E0425]: cannot find value `DIRECT_TCP_CONCURRENCY` in module `option`
  --> leaf/src/app/dispatcher.rs:92:68
   |
92 |         let (direct_tcp_tx, direct_tcp_rx) = mpsc::channel(option::DIRECT_TCP_CONCURRENCY);
   |                                                                    ^^^^^^^^^^^^^^^^^^^^^^ not found in `option`

error[E0425]: cannot find value `DNS_CACHE_SIZE` in module `option`
  --> leaf/src/app/dns_client.rs:36:21
   |
36 |             option::DNS_CACHE_SIZE,
   |                     ^^^^^^^^^^^^^^ not found in `option`

error[E0425]: cannot find value `DNS_CACHE_SIZE` in module `option`
  --> leaf/src/app/dns_client.rs:49:21
   |
49 |             option::DNS_CACHE_SIZE,
   |                     ^^^^^^^^^^^^^^ not found in `option`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0425`.
error: could not compile `leaf`
```

**Comment by mokhtarabadi** [comment]
@eycorsican I think the problem is inside the rust codes, can you help me or resolve the problem?
I trying to port leaf to android

**Comment by mokhtarabadi** [comment]
fixed by adding the android option to leaf/src/option/mod.rs

**Reply by ghost** [reply]
success run on Android device ?

**Reply by mokhtarabadi** [reply]
Yes, using cargo-ndk and android platform 21

**Reply by mokhtarabadi** [reply]
But didn't work on x86 architecture android cpus, that is tokio bug

**Reply by ghost** [reply]
> But didn't work on x86 architecture android cpus, that is tokio bug 

how to build it？

**Reply by DdVwl6Fj3a** [reply]
error: failed to run custom build command for `ring v0.16.19`  
--- stderr
  running "aarch64-linux-android-clang" "-Oz" "-DANDROID" "-ffunction-sections" "-fdata-sections" "-fPIC" "-I" "include" "-Wall" "-Wextra" "-pedantic" "-pedantic-errors" "-Wall" "-Wextra" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wundef" "-Wuninitialized" "-Wwrite-strings" "-fno-strict-aliasing" "-fvisibility=hidden" "-fstack-protector" "-g3" "-DNDEBUG" "-c" "-o/Users/apple/Documents/person/project/rust/leaf/target/aarch64-linux-android/release/build/ring-af49b8f3fbc0374e/out/aesv8-armx-linux64.o" "/Users/apple/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.19/pregenerated/aesv8-armx-linux64.S"
  thread 'main' panicked at 'failed to execute ["aarch64-linux-android-clang" "-Oz" "-DANDROID" "-ffunction-sections" "-fdata-sections" "-fPIC" "-I" "include" "-Wall" "-Wextra" "-pedantic" "-pedantic-errors" "-Wall" "-Wextra" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wundef" "-Wuninitialized" "-Wwrite-strings" "-fno-strict-aliasing" "-fvisibility=hidden" "-fstack-protector" "-g3" "-DNDEBUG" "-c" "-o/Users/apple/Documents/person/project/rust/leaf/target/aarch64-linux-android/release/build/ring-af49b8f3fbc0374e/out/aesv8-armx-linux64.o" "/Users/apple/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.19/pregenerated/aesv8-armx-linux64.S"]: No such file or directory (os error 2)', /Users/apple/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.19/build.rs:670:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed

when I build android lib, I got this error, can you give me a hand?

**Reply by ghost** [reply]
> But didn't work on x86 architecture android cpus, that is tokio bug

This is a huge improvement.Would you please tell me how to build it ?

i tried add  this code  to leaf/src/option/mod.rs,but not works

```
#[cfg(target_os = "android")]
mod ios;

#[cfg(target_os = "android")]
pub use ios::*;
```

**Reply by mokhtarabadi** [reply]
I created separate file for android options and link it in mod.rs 
Then install cargo ndk from https://github.com/bbqsrc/cargo-ndk
After that I written [this bash script](https://pastebin.com/gi9QzuLH) to build
I only need vmess support, so I build needed features

**Reply by mokhtarabadi** [reply]
Also set ANDROID_NDK_HOME environment variable to ndk 21 or 22

**Reply by DdVwl6Fj3a** [reply]
> I created separate file for android options and link it in mod.rs
> Then install cargo ndk from https://github.com/bbqsrc/cargo-ndk
> After that I written [this bash script](https://pastebin.com/gi9QzuLH) to build
> I only need vmess support, so I build needed features

can you share your separate file for android options?

**Reply by mokhtarabadi** [reply]
Same as unix.rs

**Reply by DdVwl6Fj3a** [reply]
> Same as unix.rs
Thank you very much , I'll try it

**Reply by giftedunicorn** [reply]
using api=26 can build for x86

**Reply by giftedunicorn** [reply]
leaf master will be failed for android/log.h file not found

**Reply by mokhtarabadi** [reply]
add this "-I/path_to_android_ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include" in leaf/build.rs lines 81 and 122
